Issue23

brec9824 opened the pull request:
Added follow count, the ability to follow, unfollow projects and added a popup for logging in. The login window works but I probably will have to figure out a way to edit authentication so it redirects to the page you are on. Fixes #23 I dont think we needs likes/dislikes because it seems kind of redundant. Also fixes #108 just insert the breadcrumb section above the mainBody section if you want it included. I included it with most pages but removed it from authentication views.

Commit:
Added project following

Added the ability to follow a project and unfollow a project. Also added a mini login window that can be popped up.

Commit:
Minor tweaks to database searches

Changed counting followers to server side and checking if user is already a follower or not.

Commit:
Made breadcrumb optional

Just insert the breadcrumb section in above the mainBody section if you want it included.

rickbau5 commented:
We can use the `@include`, like @backburn was doing for the file list. So if you make a `breadcrumb.blade.php` in the editor folder, you can then just do `@include('editor.breadcrumb')`. It can just be that one portion that you are copy-pasting in each file.

rickbau5 commented:
getting some wonky formatting 
![image](https://cloud.githubusercontent.com/assets/11299264/14700935/5c50b278-0757-11e6-9d10-f013f150d534.png)

rickbau5 commented:
As it stands there's no way to see all projects you're following, right? Figure that's part of the Profile stuff which is fine, just curious

rickbau5 commented on app/Http/Controllers/ProjectController.php:
```diff
@@ -25,9 +27,7 @@ public function listProjects()
-        $project->load('comments.user');
-
-        return view('project.view', ['project' => $project]);
+        return view('project.view', ['project' => $project, 'isUserFollower' => $project->isUserFollower(), 'followerCount' => $project->getFollowerCount()]);
```
Passing `isUserFollower` and `followerCount` as variables to the view seems redundant to me. `project` will be accessible in the view, so why not just do `$project->isUserFollower()` in the view itself? Idk what best practice is, but that makes more sense to me

brec9824 replied:
Yeah I wasn't to sure which way was the correct way to do this so I did it like this.

rickbau5 replied:
I would go ahead and just pass `project` and then in the view do `$project->isUserFollower()` and `$project->getFollowerCount()` 👍

rickbau5 commented on app/Project.php:
```diff
@@ -145,4 +145,84 @@ public function user()
+     *
+     * @return
+     */
+    public function isUserFollower()
```
This function signature is weird to me. I keep wanting to see a user passed as a parameter to check. If you did make it accept a user, you'd get the benefit of being able to check arbitrary users for free, which may be nice in the future. But idk how you'd handle auth in that case.

brec9824 replied:
Ok I updated it so you can check arbitrary users.

rickbau5 commented on app/Project.php:
```diff
@@ -145,4 +145,84 @@ public function user()
+     *
+     * @return
+     */
+    public function addFollower()
```
This one too, but here there's no check for auth, presumably because the user will be authenticated to see the follow button. When i try to use a direct link in incognito i get the 403 page, so that's good.

rickbau5 commented on app/Project.php:
```diff
@@ -145,4 +145,84 @@ public function user()
+    public function getFollowerIDs()
+    {
+        $followerArray = [];
+        foreach ($this->followers as $follower)
```
Think you could do

```
return array_map(function($follower) { return $follower->user_id; }, $this->followers);
```

Mapping is sweet, much nicer to write than for loops. But it's more or less the same, whichever you prefer.

brec9824 replied:
Interesting didn't know about that, but I ended up deleting the function because it actually wasn't being used.

rickbau5 commented:
I think creating a new table and model `ProjectFollower` might not be the best way to do this. `ProjectFollower` really contains no data, but only shows a relationship between a project and a user. This relationship can be defined within the project and user models themselves, no? 

When a user follows a project, it can update the `Project` model, adding that user to the list of followers as well as in the `User` model, updating the list of followed projects. 

What do you think?

rickbau5 commented:
@brec9824 reviewed. A few comments that i'd like your feedback on, see above

Commit:
Added passable user

isUserFollower can now be called with a user id to check an arbitary user.

Commit:
Fixed change in about page

Commit:
Removed uneeded function

Commit:
Updated methods with optional param

Delete follower and add follower can now be called with an optional user id. Also updated delete follower to remove the foreach.

Commit:
Updated breadcrumb

Commit:
Moved view files and delete buttons

brec9824 commented:
@rickbau5 I dont think you are supposed to create arrays on a table and to do that I believe requires serializing the array into a string. I fixed everything, now I have to figure out why travis is failing.

rickbau5 commented:
@brec9824 I don't think you committed `insersts/breadcrumb.blade.php` 😉

rickbau5 commented:
@brec9824 re: ProjectFollowers. Looks like you're right. Didn't know MySQL didn't support that. While reading up on that I found a lot of stuff on Foreign Keys, we might be wanting to use those and making our tables cascading. For example, if you delete a project it doesn't delete the comments on it. We could handle that manually in code, but I think the proper way to do it is using Foreign Keys and cascading. See [here](http://www.sitepoint.com/mysql-foreign-keys-quicker-database-development/):

rickbau5 commented on resources/views/auth/loginmini.blade.php:
```diff
@@ -0,0 +1,65 @@
+<main class="primary-main row center-block">
```
Damn I **really** like this 👍  nice thought

rickbau5 replied:
Can you `@include` this on the login page?

rickbau5 commented on resources/views/project/view.blade.php:
```diff
@@ -23,18 +22,35 @@
+                            <!-- Modal -->
+                            <div class="modal fade" id="loginForm" role="dialog">
+                                <div class="modal-dialog modal-lg">
+                                    @include('auth.loginmini')
```
Damn I **really** like this 👍  nice thought

Commit:
Added missing folder for breadcrumb

rickbau5 commented:
🎉  just two things. You can ignore the orphaned project followers for now and we can fix it along with the project comments as well. I think you can include the login form like you did for the dialog in `login.blade.php`. And then change the `view` call in the controller and we can merge it!

Commit:
Fixed view call and modified login

brec9824 commented:
@rickbau5 Ok done

rickbau5 commented:
@brec9824 sweet i'll merge once checks pass

brec9824 commented:
Nooooo travis

rickbau5 commented:
¯\_(ツ)_/¯

Commit:
Added untracked file

Added untracked login insert page and changed the name of the insert on login.

rickbau5 commented:
`View [auth.login_insert] not found. ` is the error it's failing with. It looks like the auth thing is redirecting to the wrong page?

brec9824 commented:
Found it I had an untracked file.

rickbau5 commented:
Gotta do that `git status` ;)

rickbau5 commented:
Cool, merging
